### DISCUSSION TITLE
EMQX Enterprise 4.4.16 DB Persist rule (PostgreSQL) gets stuck in Fallback condition with error "Disconnected"

### DISCUSSION THREAD

**Discussion by Jarrod848** [emqx/emqx#10330] [Q&A]
We are running EMQX Enterprise 4.4.16 in an Azure AKS cluster. We created a rule in the EMQX Rule Engine to persist telemetry messages to an Azure PostgreSQL database. This works without issue.  We then added a "Fallback Action" to redirect the messages to another table in the same database if the primary action fails for some reason (serialization, etc) . We can send an "invalid" message to the topic and trigger a SQL failure, the fallback action  is activated and the record is inserted into the DB. The issue is that all subsequent "valid" messages to the topic are routed to the fallback action with error "Disconnected".  I tried configuring the "Fallback Action" to publish the messages to another topic, and noticed the same behavior.  Is this a known issue?

Here is a snippet of the error:

`2023-04-04T19:47:31.944975+00:00 [error] user-123@10.2.68.6:443 Rule: <<"rule:995803">>; Action: data_to_pgsql; Resource: <<"resource:270378">>. Query failed reason: disconnected
2023-04-04T19:47:31.945085+00:00 [error] user-123@10.2.68.6:443 Rule: <<"rule:995803">>; Action: data_to_pgsql; Resource: <<"resource:270378">>. Continue next action, reason: disconnected

Larger attachment:
[emqx_disconnect_logs.txt](https://github.com/emqx/emqx/files/11153182/emqx_disconnect_logs.txt)
`

**Comment by kjellwinblad** [comment]
Hi, Thanks for reporting this. One possibility is that the connection to the database closes after the failure and then it could take a while for the new connection to come back. If that is the case, the bridge should have a setting called "Auto Restart Interval" or something similar that one could lower to make the bridge restart quicker. I will do some experiments to try to reproduce the issue. If we find that it is a bug, we will fix it.

**Comment by kjellwinblad** [comment]
I have tried to reproduce your issue without succeeding using the following steps:

1. Create a postgresql table: `CREATE TABLE mytab2 (myvalue VARCHAR ( 10 ));`
2. Set up a postgresql resource in EMQX
3. Set up an EMQX rule with the following content:
    ```
    SELECT
      payload.msg as msg
     FROM
      "test_in/1"
    ```
    and a postgresql action with sync mode and with the following SQL:
    ```
    INSERT INTO mytab2(myvalue)
    VALUES (${msg});
    ```
4. set up a fallback action that republish to another topic
5. send message to `test_in/1` with payload {msg:"hello"}
6. a row with hello gets stored in the table
7. send message to `test_in/1` with payload {msg:"hellohellohello"}
8. This fails because the value is to big for the db column
9. The fallback topic gets a message
10. send message to `test_in/1` with payload {msg:"hello2"}
11. the fallback action is not triggered and the database gets a new row with hello2 which is what one would expect

@Jarrod848 we would be very thankful if you could send us more details about the configuration you use to get the issue so we can reproduce it? For example, what type of invalid message to you send?

**Reply by kjellwinblad** [reply]
Looking at the log, it looks like our EMQX's PostgreSQL driver crashes due to the invalid SQL that is given to it and this makes the connection go down. The connection should start again after some time but if this is not happening it is a bug. It would be great if you can send us an example of the invalid SQL that trigger this issue as it will make it faster for us to reproduce and fix the issue.

**Reply by kjellwinblad** [reply]
I created an issue based on this discussion https://github.com/emqx/emqx/issues/10342 as I think this is most likely a bug. We can use this issue for further discussion and to track the progress of this.

**Reply by Jarrod848** [reply]
Appreciate the response.  It looks like I can trigger the failure when mapping a json timestamp value to a Postgres "timestamptz" (timestamp with timezone) column. 

Create Table:
`CREATE TABLE mytab3 (myvalue VARCHAR (100), ts timestamptz);`

EMQX Rule:
`SELECT
  payload.msg as msg,
  payload.ts as ts
 FROM
  "test/1"`

Postgres Action:
`INSERT INTO mytab3(myvalue, ts)
VALUES (${msg},${ts});`

Send Json:
`{ "msg": "hello", "ts": "2023-04-06 15:23:59.690549+00"}`

Error:
`2023-04-07T16:51:54.894151+00:00 [error] Generic server <0.4180.0> terminating. Reason: {function_clause,[{epgsql_idatetime,timestamp2i,[<<"2023-04-07 12:50:12.176 -0400">>],[{file,"epgsql_idatetime.erl"},{line,97}]},{epgsql_idatetime,encode,2,[{file,"epgsql_idatetime.erl"}`

Update Postgres Action to use "now()":
`INSERT INTO mytab3(myvalue, ts)
VALUES (${msg},now());`

Send Json:
`{ "msg": "hello", "ts": "2023-04-06 15:23:59.690549+00"}`

Success:
No error, record is inserted into the db successfully.  It seems to be throwing the error when parsing the timestamp value: ${ts} . I did verify that I can insert this value into the Postgres table without error.

**Reply by kjellwinblad** [reply]
Thanks, this will be useful when tracking down this issue!

**Reply by TimisRobert** [reply]
I'm getting the same error using the EMQX Enterprise docker, latest version. It's weird because the same expression used in EMQX cloud works without issues.
